server/venueless/core/models/world.py

```diff
@@ -44,10 +80,68 @@ def decode_token(self, token):
                     issuer=issuer,
                 )
 
-    def has_permission(self, permission, traits, extra_config=None):
-        permission_config = copy.deepcopy(self.permission_config)
-        if extra_config:
-            permission_config.update(extra_config)
-        return permission in permission_config and all(
-            trait in traits for trait in permission_config[permission]
-        )
+    def has_permission_implicit(self, *, traits, permission: Permission, room=None):
+        for role, required_traits in self.trait_grants.items():
+            if isinstance(required_traits, list) and all(
+                r in traits for r in required_traits
+            ):
+                if permission.value in self.roles[role]:
+                    return True
+
+        if room:
+            for role, required_traits in room.trait_grants.items():
+                if isinstance(required_traits, list) and all(
+                    r in traits for r in required_traits
+                ):
+                    if permission.value in self.roles[role]:
+                        return True
+
+    def has_permission(self, *, user, permission: Permission, room=None):
+        """
+        Returns whether a user holds a given permission either on the world or on a specific room.
+        """
+        if self.has_permission_implicit(
+            traits=user.traits, permission=permission, room=room
+        ):
+            return True
+
+        roles = user.get_role_grants(room)
+        for r in roles:
+            if permission.value in self.roles[r]:
+                return True
+
+    async def has_permission_async(self, *, user, permission: Permission, room=None):
+        """
+        Returns whether a user holds a given permission either on the world or on a specific room.
+        """
+        if self.has_permission_implicit(
+            traits=user.traits, permission=permission, room=room
+        ):
+            return True
+
+        roles = await database_sync_to_async(user.get_role_grants)(room)
+        for r in roles:
+            if permission.value in self.roles[r]:
+                return True
+
+    def get_all_permissions(self, user):
+        result = defaultdict(set)
+        for role, required_traits in self.trait_grants.items():
+            if isinstance(required_traits, list) and all(
+                r in user.traits for r in required_traits
+            ):
+                result[self].update(self.roles[role])
+
+        for grant in user.world_grants.all():
+            result[self].update(self.roles[grant.role])
+
+        for room in self.rooms.all():
```

Comment:
Not safe in the meaning of "can't be called from an async thread"? Hmm yeah. I figured inside a model is a place where you should expect some database access, but maybe I'm wrong.

I want to do a round of refactoring of the codebase after this PR is merged. Even if it will lead to duplicate code, my end goal would be that ideally ``database_sync_to_async`` is called at most once per websocket event, since I believe these thread context switches are pretty expensive performance-wise. Not sure if that goal is feasible, but I'll try (if you agree with the objective)